compiler/hash-typecheck/src/new/passes/symbol_resolution.rs

```diff
@@ -120,8 +276,644 @@ impl<'tc> SymbolResolutionPass<'tc> {
+        &self,
+        component: &AstPathComponent<'_>,
+        starting_from: Option<(ModMemberValue, Span)>,
+        in_expr: InExpr,
+        total_span: Span,
+    ) -> TcResult<ResolvedAstPathComponent> {
+        let binding = self.resolve_ast_name(component.name, component.name_span, starting_from)?;
+
+        match binding.kind {
+            BindingKind::ModMember(_, mod_member_id) => {
+                let mod_member = self.stores().mod_members().get_element(mod_member_id);
+                match mod_member.value {
+                    ModMemberValue::Data(data_def_id) => {
+                        let data_def_params =
+                            self.stores().data_def().map_fast(data_def_id, |def| def.params);
+                        let args =
+                            self.apply_ast_args_to_def_params(data_def_params, &component.args)?;
+                        Ok(ResolvedAstPathComponent::Data(data_def_id, args))
+                    }
+                    ModMemberValue::Mod(mod_def_id) => {
+                        let mod_def_params =
+                            self.stores().mod_def().map_fast(mod_def_id, |def| def.params);
+                        let args =
+                            self.apply_ast_args_to_def_params(mod_def_params, &component.args)?;
+                        Ok(ResolvedAstPathComponent::Mod(mod_def_id, args))
+                    }
+                    ModMemberValue::Fn(fn_def_id) => {
+                        // If a function is used in a value position, then it is
+                        // the function's address that is used.
+                        if in_expr == InExpr::Ty {
+                            return Err(TcError::CannotUseFunctionInTypePosition {
+                                location: self.source_location(total_span),
+                            });
+                        }
+
+                        // Apply any arguments to the function.
+                        let resultant_term = self.wrap_term_in_fn_call_from_ast_args(
+                            self.new_term(Term::FnRef(fn_def_id)),
+                            &component.args,
+                        );
+
+                        // Return a term or a type as appropriate by `in_expr`.
+                        match in_expr {
+                            InExpr::Ty => Ok(ResolvedAstPathComponent::Ty(
+                                self.new_ty(Ty::Eval(resultant_term)),
+                            )),
+                            InExpr::Value => Ok(ResolvedAstPathComponent::Term(resultant_term)),
+                        }
+                    }
+                }
+            }
+            BindingKind::Ctor(_, ctor_def_id) => {
+                // A constructor cannot be namespaced further, so it becomes
+                // a term.
+                if in_expr == InExpr::Ty {
+                    // @@Improvement: support shorthand refinement like `None` as `Option<T> of
+                    // None`.
```

Comment:
I think even rust hacks this by just doing `use Option::*` in the prelude

Reply:
Yeah, this is just to note that this should be supported.

Reply:
It might be worth introducing a similar language feature like `use` which can bring members of data types into scope, but of course not now